File: 0042-trapping-rain-water/0042-trapping-rain-water.go
type rec struct {
    val, idx int
}

func computeNGE(nums []int, backward bool) []int {
    n := len(nums)
    stack := make([]rec, 0)
    nge := make([]int, n)

    if !backward {
        for r:=0;r<n;r++{
            for len(stack) > 0 && stack[len(stack)-1].val <= nums[r] {
                pop := stack[len(stack)-1]
                nge[pop.idx] = r //point out the nge of item idx is at r index
                stack = stack[0:len(stack)-1] // pop it
            }

            stack = append(stack, rec{nums[r], r})
        }

        for len(stack) > 0 {
            pop := stack[len(stack)-1]
            nge[pop.idx] = -1 // no greater element
            stack = stack[0:len(stack)-1] // pop it
        }

    }else{
        for r:=n-1;r>=0;r--{
            for len(stack) > 0 && stack[len(stack)-1].val <= nums[r] {
                pop := stack[len(stack)-1]
                nge[pop.idx] = r //point out the nge of item idx is at r index
                stack = stack[0:len(stack)-1] // pop it
            }

            stack = append(stack, rec{nums[r], r})
        }

        for len(stack) > 0 {
            pop := stack[len(stack)-1]
            nge[pop.idx] = -1 // no greater element
            stack = stack[0:len(stack)-1] // pop it
        }
    }
    return nge
}
func trap(heights []int) int {
    n := len(heights)
    if n <= 2 {
        return 0 // at least 3 lengths of bar can form the concave
    }
    // the idea is to find the next and prev greater element, so it forms the concave that can trap the water
    nge := computeNGE(heights, false)
    pge := computeNGE(heights, true)

    water := 0
    // fmt.Println(heights)
    // fmt.Println("pge", pge)
    // fmt.Println("nge", pge)
    i:=1
    for i < n {
        subtotal := 0
        
        l := i
        // search the highest in the left, while still maintaining height constraint
        for pge[l] != -1 && nge[l]!=-1{
            l = pge[l]
        }
    
        r := i
        // search the highest in the right, while still maintaining height constraint
        for nge[r] != -1 && pge[r] != -1{
            r = nge[r]
        }

        if i != l && i !=r {
            // fmt.Println("concave", l, i, r)
            h := min(heights[l], heights[r])
            for l+1 < r {
                l++
                diff := h - heights[l]
                if diff >0{
                    subtotal += diff
                }
            }
            i = r+1
        }else{
            i++
        }
        water += subtotal

    }
    return water
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.rs
use std::vec::Vec;
impl Solution {
    fn next_greater_element(arr: &Vec<i32>, is_forward: bool) -> Vec<i32> {
        let mut nge: Vec<i32> = vec![-1;arr.len()];
        let mut stack: Vec<usize> = vec![];
        if is_forward {

            for (pos, &el) in arr.iter().enumerate() {
                while stack.len() > 0 && arr[*stack.last().unwrap()] < el {
                    let last = stack.pop().unwrap();
                    nge[last] = pos as i32;
                }
                stack.push(pos);
            }

        }else{
            for (pos, &el) in arr.iter().enumerate().rev() {
                while stack.len() > 0 && arr[*stack.last().unwrap()] < el {
                    let last = stack.pop().unwrap();
                    nge[last] = pos as i32;
                }
                stack.push(pos);
            }
        }

        while stack.len() > 0 {
            nge[stack.pop().unwrap()] = -1
        }

        return nge;
    }

    pub fn trap(height: Vec<i32>) -> i32 {
        let nge = Solution::next_greater_element(&height, true);
        let pge = Solution::next_greater_element(&height, false);
      
        let mut i = 0;
        let mut total_water = 0;
        while i < height.len() {
            let mut right = i;
            while nge[right] != -1 && pge[right] != -1 {
                right = nge[right] as usize;
            }

            let mut left = i;
            while pge[left] != -1 && nge[left] != -1 {
                left = pge[left] as usize;
            }

            let min = std::cmp::min(height[left], height[right]);
            if left != i && right != i {
                for pos in left+1..right {
                    total_water += min - height[pos];
                }
                i = right + 1;
            } else{
                i+=1;
            }

        }
        return total_water

    }
}